src/trace_segment.ml

```diff
@@ -0,0 +1,799 @@
+        [frame]. *)
+    val become_frame : t -> Location.t -> parent:frame -> frame
+  end
+
+  module For_testing : sig
+    val to_string_list : t -> string list
+    val print_callstack : t -> unit
+  end
+end = struct
+  type t =
+    { mutable location : Event.Location.t
+    ; mutable parent : t Or_null.t
+    }
+
+  let[@inline always] create location ~parent = { location; parent }
+
+  let rec root t =
+    match t with
+    | { parent = Null; _ } | { parent = This { parent = Null; _ }; _ } -> t
+    | { parent = This parent; _ } -> root parent
+  ;;
+
+  let rec find t target =
+    match t with
+    | { parent = Null; _ } -> Null
+    | { location = { symbol; _ }; _ } when Symbol.equal symbol target -> This t
+    | { parent = This parent; _ } -> find parent target
+  ;;
+
+  let rec iter_until t stop_symbol ~f =
+    match t with
+    | { parent = Null; _ } -> ()
+    | { location = { symbol; _ }; _ } when Symbol.equal symbol stop_symbol -> ()
+    | { parent = This parent; _ } ->
+      f t;
+      iter_until parent stop_symbol ~f
+  ;;
+
+  let rec iter_until_rev t stop_symbol ~f =
+    match t with
+    | { parent = Null; _ } -> ()
+    | { location = { symbol; _ }; _ } when Symbol.equal symbol stop_symbol -> ()
+    | { parent = This parent; _ } ->
+      iter_until_rev parent stop_symbol ~f;
+      f t
+  ;;
+
+  module Sentinel = struct
+    type nonrec t = t
+
+    let sentinel_location : Location.t =
+      { instruction_pointer = 0L; symbol_offset = 0; symbol = Unknown }
+    ;;
+
+    let[@inline always] create () = { location = sentinel_location; parent = Null }
+
+    let become_frame t location ~parent =
+      t.location <- location;
+      t.parent <- This parent;
+      t
+    ;;
+  end
+
+  module For_testing = struct
+    let rec to_string_list acc t =
+      match t.parent with
+      | Null -> acc
+      | This parent ->
+        to_string_list (Symbol.display_name t.location.symbol :: acc) parent
+    ;;
+
+    let to_string_list t = to_string_list [] t
+    let print_callstack leaf = to_string_list leaf |> String.concat_lines |> print_endline
+  end
+end
+
+module Control_flow = struct
+  type t =
+    | Jump
+    | Call
+    | Return
+end
+
+module Callstack = struct
+  type t =
+    #{ time : Timestamp.t
+     ; leaf : Frame.t
+     ; control_flow : Control_flow.t
+     }
+end
+
+type t =
+  { mutable root : Frame.Sentinel.t
+  ; callstacks : Callstack.t Nonempty_vec.t
+  }
+
+let create () =
+  let root = Frame.Sentinel.create () in
+  { root
+  ; callstacks =
+      Nonempty_vec.create
+        (#{ time = Timestamp.zero; leaf = (root :> Frame.t); control_flow = Return }
+         : Callstack.t)
+  }
+;;
+
+let[@inline always] current_frame t = (Nonempty_vec.last t.callstacks).#leaf
+
+let replace_root t location =
+  let new_sentinel = Frame.Sentinel.create () in
+  let root =
+    Frame.Sentinel.become_frame t.root location ~parent:(new_sentinel :> Frame.t)
```

Comment:
Ah yeah, if we typed `parent:t` here, we could avoid the coercion.

Reply:
Addressed by [comment above](https://github.com/janestreet/magic-trace/pull/350/files/9bbeaa5d48d5b33121b91f2abbf877d55014e2c2#r2636192365)